File: x-fly/src/main/java/com/xia/fly/ui/dialog/loader/LoaderCreator.java
package com.xia.fly.ui.dialog.loader;

import com.wang.avi.AVLoadingIndicatorView;
import com.wang.avi.Indicator;

import java.util.WeakHashMap;

/**
 * @author xia
 * @date 2018/7/29.
 */
@SuppressWarnings("UnusedReturnValue")
public final class LoaderCreator {
    private static final WeakHashMap<String, Indicator> LOADING_MAP = new WeakHashMap<>();

    public static AVLoadingIndicatorView create(String type, AVLoadingIndicatorView avLoadingIndicatorView) {
        if (LOADING_MAP.get(type) == null) {
            final Indicator indicator = getIndicator(type);
            LOADING_MAP.put(type, indicator);
        }
        avLoadingIndicatorView.setIndicator(LOADING_MAP.get(type));
        return avLoadingIndicatorView;
    }

    private static Indicator getIndicator(String name) {
        if (name == null || name.isEmpty()) {
            return null;
        }
        final StringBuilder drawableClassName = new StringBuilder();
        if (!name.contains(".")) {
            final String defaultPackageName = AVLoadingIndicatorView.class.getPackage().getName();
            drawableClassName.append(defaultPackageName)
                    .append(".indicators")
                    .append(".");
        }
        drawableClassName.append(name);
        try {
            final Class<?> drawableClass = Class.forName(drawableClassName.toString());
            return (Indicator) drawableClass.newInstance();
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: fly/src/main/java/com/xia/fly/ui/dialog/loader/LoaderCreator.kt
package com.xia.fly.ui.dialog.loader

import com.wang.avi.AVLoadingIndicatorView
import com.wang.avi.Indicator
import java.util.*

/**
 * @author xia
 * @date 2018/7/29.
 */
object LoaderCreator {

    private val LOADING_MAP = WeakHashMap<String, Indicator>()

    fun create(@LoaderStyles.LoaderStyle type: String,
               indicatorView: AVLoadingIndicatorView): AVLoadingIndicatorView {
        var indicator = LOADING_MAP[type]
        if (indicator == null) {
            indicator = getIndicator(type)
            LOADING_MAP[type] = indicator
        }
        indicatorView.indicator = indicator
        return indicatorView
    }

    private fun getIndicator(name: String?): Indicator? {
        if (name == null || name.isEmpty()) {
            return null
        }
        val drawableClassName = StringBuilder()
        if (!name.contains(".")) {
            val defaultPackageName = AVLoadingIndicatorView::class.java.getPackage()!!.name
            drawableClassName.append(defaultPackageName)
                    .append(".indicators")
                    .append(".")
        }
        drawableClassName.append(name)
        return try {
            val drawableClass = Class.forName(drawableClassName.toString())
            drawableClass.newInstance() as Indicator
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }
}
